### PR TITLE
[Kernel] Add FP8 KV cache support to Triton MLA decode attention

### PR DESCRIPTION
**Pull request by grimulkan**
Enable fp8/fp8_e4m3 KV cache for the Triton MLA attention backend, which is the only MLA backend available on sm120 GPUs.

- Add fp8 and fp8_e4m3 to TritonMLABackend.supported_kv_cache_dtypes
- Thread k_scale/v_scale through decode attention kernel launch path
- Add FP8 dequant-on-load in both stage1 Triton kernels (MHA and grouped/MLA)
- Set supports_quant_query_input=False for FP8 (BF16 queries + FP8 KV)
- Add FP8-specific parametrized test cases

## Purpose
Enable FP8 KV cache for MLA models on sm120 (Blackwell consumer GPUs). The Triton MLA backend is the only available MLA backend on sm120, but previously blocked FP8 with `NotImplementedError`.

## Changes
- Add `"fp8"` and `"fp8_e4m3"` to `TritonMLABackend.supported_kv_cache_dtypes`
- Thread `k_scale`/`v_scale` through the decode attention kernel launch path
- Add FP8 dequant-on-load pattern in both Triton kernels (grouped and non-grouped paths)
- Set `supports_quant_query_input=False` for (BF16 queries + FP8 KV cache)
- Add 16 FP8-specific parametrized test cases

## Test Plan
- Run `pytest tests/kernels/attention/test_triton_decode_attention.py`
- Test inference using `vllm serve` on a compatible model using Triton MLA backend
- End-to-end lm_eval (GSM8K)

## Test Results
- All 112 tests pass (96 existing BF16 + 16 new FP8) in `tests/kernels/attention/test_triton_decode_attention.py`
- Tested on 16 x RTX 6000 Pro GPUs with Kimi K2.5 with DCP = 1 and fp8 KV cache
```
local-completions ({'model': 'Kimi-K2.5', 'tokenizer_backend': '/models/moonshotai-Kimi-K2.5', 'base_url': 'http://localhost:8000/v1/completions', 'num_concurrent': 500, 'tokenized_requests': False}), gen_kwargs: ({}), limit: None, num_fewshot: 5, batch_size: 1

--attention-backend TRITON_MLA --kv-cache-dtype auto

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9378|±  |0.0067|
|     |       |strict-match    |     5|exact_match|↑  |0.9378|±  |0.0067|

--attention-backend TRITON_MLA --kv-cache-dtype fp8

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9393|±  |0.0066|
|     |       |strict-match    |     5|exact_match|↑  |0.9393|±  |0.0066|
```
~0.15 pts within expected tolerance. Normalized generation speed (ignoring the potential 2x higher concurrency with fp8) is about the same as bf16, which is to be expected in this approach.

## Known limitations
- DCP (decode context parallelism) + FP8 is not supported (pre-existing limitation in MLA common layer, affects all backends) See https://github.com/vllm-project/vllm/pull/34795 for a fix.
- To keep the edits simple, this approach saves memory but the compute is still in bf16. It is not a true fp8 tensor core op.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by pavanimajety** [comment]
@grimulkan thanks for your contribution! Could you please add end to end lm_eval results for any of the standard models that can run on SM120?

**Comment by voipmonitor** [comment]
I'm confirming that this is working on 8x RTX PRO AMD Turin:

NCCL_P2P_LEVEL=SYS  VLLM_LOG_STATS_INTERVAL=1  NCCL_GRAPH_FILE=/mnt/nccl_graph_opt.xml  VLLM_TEST_FORCE_FP8_MARLIN=1 VLLM_MARLIN_USE_ATOMIC_ADD=1 VLARLIN_INPUT_DTYPE=fp8 vllm serve moonshotai/Kimi-K2.5 --served-model-name Kimi-K2.5 --trust-remote-code --host 0.0.0.0 --port 5000 --tensor-parallel-size 8 --pipeline-parallel-size 1 --enable-chunked-prefill --enable-prefix-caching --load-format fastsafetensors --tool-call-parser kimi_k2 --enable-auto-tool-choice --reasoning-parser kimi_k2 --async-scheduling --gpu-memory-utilization 0.93 --max-num-batched-tokens 4096 --mm-processor-cache-gb 0 --mm-encoder-tp-mode weights --language-model-only --attention-backend TRITON_MLA --kv-cache-dtype fp8 

GPU KV cache size: 449,600 tokens
speed: 79tok/sec 

when --decode-context-parallel-size 8 is used (more KV cache): 
GPU KV cache size: 3,621,504 tokens

speed: 66tok/sec

**Comment by grimulkan** [comment]
Rebased (only documentation conflict)

File: vllm/v1/attention/ops/triton_decode_attention.py [hunk 1]
```diff
@@ -129,6 +132,8 @@ def _fwd_kernel_stage1(
                 mask=(offs_n[:, None] < split_kv_end) & (mask_d[None, :]),
                 other=0.0,
             )
+            if k.dtype.is_fp8():
+                k = (k.to(tl.float32) * tl.load(k_scale)).to(q.dtype)
```
Comment (MatthewBonanni) [inline vllm/v1/attention/ops/triton_decode_attention.py]: We should move the `tl.load(k_scale)` outside the loop. Not really a big deal though

File: vllm/v1/attention/ops/triton_decode_attention.py [hunk 2]
```diff
@@ -147,6 +152,8 @@ def _fwd_kernel_stage1(
                 mask=(offs_n[:, None] < split_kv_end) & (mask_dv[None, :]),
                 other=0.0,
             )
+            if v.dtype.is_fp8():
+                v = (v.to(tl.float32) * tl.load(v_scale)).to(q.dtype)
```
Comment (MatthewBonanni) [inline vllm/v1/attention/ops/triton_decode_attention.py]: ditto

File: vllm/v1/attention/ops/triton_decode_attention.py [hunk 3]
```diff
@@ -336,6 +349,8 @@ def _fwd_grouped_kernel_stage1(
                 mask=(offs_n[None, :] < split_kv_end) & (mask_d[:, None]),
                 other=0.0,
             )
+            if k.dtype.is_fp8():
+                k = (k.to(tl.float32) * tl.load(k_scale)).to(q.dtype)
```
Comment (MatthewBonanni) [inline vllm/v1/attention/ops/triton_decode_attention.py]: ditto

File: vllm/v1/attention/ops/triton_decode_attention.py [hunk 4]
```diff
@@ -348,6 +363,8 @@ def _fwd_grouped_kernel_stage1(
                     mask=(offs_n[None, :] < split_kv_end) & (mask_dpe[:, None]),
                     other=0.0,
                 )
+                if kpe.dtype.is_fp8():
+                    kpe = (kpe.to(tl.float32) * tl.load(k_scale)).to(qpe.dtype)
```
Comment (MatthewBonanni) [inline vllm/v1/attention/ops/triton_decode_attention.py]: ditto

File: vllm/v1/attention/ops/triton_decode_attention.py [hunk 5]
```diff
@@ -368,6 +385,8 @@ def _fwd_grouped_kernel_stage1(
                 mask=(offs_n[:, None] < split_kv_end) & (mask_dv[None, :]),
                 other=0.0,
             )
+            if v.dtype.is_fp8():
+                v = (v.to(tl.float32) * tl.load(v_scale)).to(q.dtype)
```
Comment (MatthewBonanni) [inline vllm/v1/attention/ops/triton_decode_attention.py]: ditto

File: vllm/v1/attention/ops/triton_decode_attention.py [hunk 6]
```diff
@@ -671,8 +702,21 @@ def decode_attention_fwd(
     sm_scale,
     page_size=1,
     logit_cap=0.0,
+    k_scale=None,
+    v_scale=None,
 ):
     assert num_kv_splits == attn_logits.shape[2]
+
+    # When k_scale/v_scale are None (e.g. standalone / test usage without
```
Comment (MatthewBonanni) [inline vllm/v1/attention/ops/triton_decode_attention.py]: nit: this comment isn't necessary, you can remove it

File: tests/kernels/attention/test_triton_decode_attention.py [hunk 7]
```diff
@@ -90,3 +90,139 @@ def test_decode_attention(B, L, H_Q, H_KV, D_QK, D_V, CACHE_SIZE, PAGE_SIZE):
...
+            sm_scale=sm_scale,
+            page_size=PAGE_SIZE,
+            k_scale=k_scale,
+            v_scale=v_scale,
+        )
+
+    # FP8 has ~3 mantissa bits, so expect some quantization noise.
+    # Use slightly generous tolerances since random data can produce rare
+    # outliers from FP8 rounding (flashinfer uses atol=1e-2, rtol=2e-1).
+    torch.testing.assert_close(o_ref, o_fp8, atol=2e-2, rtol=3e-1)
```
Comment (MatthewBonanni) [inline tests/kernels/attention/test_triton_decode_attention.py]: Is there a reason this rtol is so loose? Other backends use 1e-2

**Review by MatthewBonanni — COMMENTED** [review]
Overall looks good, thanks for doing this! Just a few small comments

File: tests/kernels/attention/test_triton_decode_attention.py (hunk 7 above)
Comment (grimulkan) [reply tests/kernels/attention/test_triton_decode_attention.py]: Good question. I set it here to avoid some issues and chalked it up to fp8 since I couldn't find anything obviously wrong. Let me try again and see if I can drop it to what flashinfer or any other fp8 attention backend uses at least.

File: vllm/v1/attention/ops/triton_decode_attention.py (hunk 1 above)
Comment (grimulkan) [reply vllm/v1/attention/ops/triton_decode_attention.py]: Moved. I was sloppy because it's a compile time check.

File: vllm/v1/attention/ops/triton_decode_attention.py (hunk 2 above)
Comment (grimulkan) [reply vllm/v1/attention/ops/triton_decode_attention.py]: Fixed

File: vllm/v1/attention/ops/triton_decode_attention.py (hunk 3 above)
Comment (grimulkan) [reply vllm/v1/attention/ops/triton_decode_attention.py]: Fixed

File: vllm/v1/attention/ops/triton_decode_attention.py (hunk 4 above)
Comment (grimulkan) [reply vllm/v1/attention/ops/triton_decode_attention.py]: Fixed

File: vllm/v1/attention/ops/triton_decode_attention.py (hunk 5 above)
Comment (grimulkan) [reply vllm/v1/attention/ops/triton_decode_attention.py]: Fixed

File: vllm/v1/attention/ops/triton_decode_attention.py (hunk 6 above)
Comment (grimulkan) [reply vllm/v1/attention/ops/triton_decode_attention.py]: Done

File: tests/kernels/attention/test_triton_decode_attention.py (hunk 7 above)
Comment (grimulkan) [reply tests/kernels/attention/test_triton_decode_attention.py]: The other backends indeed use rtol=1e-2 but use a higher atol, so I was actually using a tighter overall tolerance (like flashinfer upstream). I changed it to match other fp8 backends in vllm. It passes over multiple tries.

**Review by MatthewBonanni — APPROVED** [review]
LGTM

**Comment by andyluo7** [comment]
## ✅ ROCm MI300X Verification — All Tests Pass

Tested this PR on **AMD Instinct MI300X (gfx942, ROCm 7.0.2)** by patching into `vllm/vllm-openai-rocm:v0.17.1`.

### Unit Tests: 16/16 PASSED ✅
```
test_decode_attention_fp8[1-16384-128-128-32-32-1025-3]   PASSED
test_decode_attention_fp8[1-16384-128-128-8-32-1025-3]    PASSED
test_decode_attention_fp8[1-16384-128-576-32-32-1025-3]   PASSED
test_decode_attention_fp8[1-16384-128-576-8-32-1025-3]    PASSED
test_decode_attention_fp8[1-16384-512-128-32-32-1025-3]   PASSED
test_decode_attention_fp8[1-16384-512-128-8-32-1025-3]    PASSED
test_decode_attention_fp8[1-16384-512-576-32-32-1025-3]   PASSED
test_decode_attention_fp8[1-16384-512-576-8-32-1025-3]    PASSED
test_decode_attention_fp8[16-16384-128-128-32-32-1025-3]  PASSED
test_decode_attention_fp8[16-16384-128-128-8-32-1025-3]   PASSED
test_decode_attention_fp8[16-16384-128-576-32-32-1025-3]  PASSED
test_decode_attention_fp8[16-16384-128-576-8-32-1025-3]   PASSED
test_decode_attention_fp8[16-16384-512-128-32-32-1025-3]  PASSED
test_decode_attention_fp8[16-16384-512-128-8-32-1025-3]   PASSED
test_decode_attention_fp8[16-16384-512-576-32-32-1025-3]  PASSED
test_decode_attention_fp8[16-16384-512-576-8-32-1025-3]   PASSED

16 passed in 12.81s
```

### E2E Serving: DeepSeek-V2-Lite (MLA) with FP8 KV cache ✅
- **Config:** `--attention-backend TRITON_MLA --kv-cache-dtype fp8 --tensor-parallel-size 1`
- **Throughput:** ~72 tok/s @ concurrency=1 (256 output tokens)
- **KV cache capacity:** 9.59M tokens (vs ~4.8M BF16 — ~2x memory savings)
- **Output quality:** Correct, coherent generation confirmed

### ROCm Notes
- Kernel code is fully cross-platform — no changes needed for ROCm
- `k.dtype.is_fp8()` correctly handles both `float8_e4m3fn` (CUDA) and `float8_e4m3fnuz` (ROCm)
- FP8 dequant via simple float32 multiply works identically on both platforms
- No `tl.range()` or other ROCm-incompatible Triton ops

Great work @grimulkan! 🎉
